packages/internal/src/storage-test-utils.js

```diff
@@ -88,15 +88,66 @@ const makeSlotStringUnserialize = () => {
 };
 export const slotStringUnserialize = makeSlotStringUnserialize();
 
+/**
+ * @example to iterate over storage messages
+ *
+ * ```js
+ * const q = makeAsyncQueue<StorageMessage>();
+ * const storage = makeFakeStorageKit('published', undefined, { eachMessage: q.enqueue });
+ * for (const message of q.iterable) { ... }
+ * ```
+ *
+ * @template T
+ */
+export const makeAsyncQueue = () => {
+  /** @type {T[]} */
+  const queue = [];
+  /** @type {null | ((r: undefined) => void)} */
+  let resolve = null;
+  let done = false;
+
+  /** @param {T} data */
+  const enqueue = data => {
+    queue.push(data);
+    if (resolve) {
+      resolve(undefined);
+      resolve = null;
+    }
+  };
+
+  const iterable = {
+    async *[Symbol.asyncIterator]() {
+      await null;
+      while (!done) {
```

Comment:
Since the iterator destructively consumes the queue, I'd suggest asserting that it isn't being reused multiple times (which would cause arbitrary interleaving). Something like:

```suggestion
  let alreadyIterating = false;
  const iterable = {
    async *[Symbol.asyncIterator]() {
      assert(!alreadyIterating, `AsyncQueue is already iterating`);
      alreadyIterating = true;
      await null;
      while (!done) {
```

Reply:
yes, this came up during development. such a diagnostic would have helped